app/code/Magento/MediaGalleryUi/Model/Directories/GetDirectoryTree.php

```diff
@@ -48,81 +57,61 @@ public function __construct(
      */
     public function execute(): array
     {
-        $tree = [
-            'name' => 'root',
-            'path' => '/',
-            'children' => []
-        ];
-        $directories = $this->getDirectories();
-        foreach ($directories as $idx => &$node) {
-            $node['children'] = [];
-            $result = $this->findParent($node, $tree);
-            $parent = &$result['treeNode'];
-
-            $parent['children'][] = &$directories[$idx];
-        }
-        return $tree['children'];
+        return $this->getDirectories();
     }
 
     /**
-     * Build directory tree array in format for jstree strandart
+     * Read media directories recursively and build directory tree array in the jstree format
      *
+     * @param string $path
      * @return array
      * @throws ValidatorException
      */
-    private function getDirectories(): array
+    private function getDirectories(string $path = ''): array
     {
         $directories = [];
 
-        /** @var Read $directory */
-        $directory = $this->filesystem->getDirectoryRead(DirectoryList::MEDIA);
-
-        if (!$directory->isDirectory()) {
-            return $directories;
-        }
-
-        foreach ($directory->readRecursively() as $path) {
-            if (!$directory->isDirectory($path) || $this->isPathExcluded->execute($path)) {
-                continue;
+        $absolutePath = $this->getMediaDirectory()->getAbsolutePath($path);
+        foreach ($this->glob->glob(rtrim($absolutePath, '/') . '/*', Glob::GLOB_ONLYDIR) as $childPath) {
```

Comment:
The `Glob()` class won't work with remote storages, all FS operations must go directly thought `DriverInterface`

Reply:
@shiftedreality so how we can read only directories with DriverInterface ? there is no any filter for that, iterator will go over all files for example if we have 1 million files and one directory that we need to get, it will takes 30 min with DriverInterface, instead of 0.1 sec.

Reply:
There is clearly missing a param in `search()`. I would suggest adding an optional pattern parameter to pass into the `glob`

Reply:
@shiftedreality ok

Reply:
> There is clearly missing a param in `search()`. I would suggest adding an optional pattern parameter to pass into the `glob`
Hello, @shiftedreality. Unfortunately, changing the interface method declaration is a backward-incompatible change, so we should avoid it.

Reply:
Definitely, but we can do it within the next release 2.5.0, once the branch is created

Reply:
@shiftedreality  that's a lot of time, the main idea to allow users such as @perfpcs use the New Media Gallery in the next release.

Reply:
@Nazar65 , I really wish I could use the new media gallery but even with the file changes, it still does not work properly, nor does the Adobe Stock sign in. I had to switch back to the old media gallery just to be able to get anything done.